yjit/src/codegen.rs

```diff
@@ -5893,6 +5893,95 @@ fn jit_rb_str_empty_p(
+
+    // Ensure the codepoint argument is a Fixnum.
+    let arg = asm.stack_opnd(0);
+    let comptime_arg = jit.peek_at_stack(&asm.ctx, 0);
+    if comptime_arg.fixnum_p() {
+        jit_guard_known_klass(
+            jit,
+            asm,
+            ocb,
+            comptime_arg.class_of(),
+            arg,
+            arg.into(),
+            comptime_arg,
+            SEND_MAX_DEPTH,
+            Counter::guard_send_not_fixnums,
+        );
+    } else if ! unsafe { RB_TYPE_P(comptime_arg, RUBY_T_STRING) } {
+        return jit_rb_str_concat(jit, asm, ocb, _ci, _cme, _block, _argc, _known_recv_class);
+    } else {
+        return false;
+    }
+
+    // Either of the string concatenation functions we call will reallocate the string to grow its
+    // capacity if necessary. In extremely rare cases (i.e., string exceeds `LONG_MAX` bytes),
+    // either of the called functions will raise an exception.
+    jit_prepare_non_leaf_call(jit, asm);
+
+    let codepoint = asm.stack_pop(1);
+    let recv = asm.stack_pop(1);
+
+    // In order to use the fast path (rb_str_buf_cat_byte), the string encoding must be ASCII-8BIT
+    // and the codepoint must be in the byte range (0x00 - 0xff).
+    // If either of those conditions are not met we must use the general string concat (str_buf_cat)
+    // function with the original codepoint argument.
+
+    let generic_str_concat_codepoint = asm.new_label("generic_str_concat_codepoint");
+    let ret_label = asm.new_label("jit_rb_str_concat_codepoint_return");
+
+    // Check if the string is ASCII-8BIT. If it isn't, we need to use the generic string concatenation.
+    asm_comment!(asm, "Check if string is ASCII-8BIT");
+    let recv_reg = asm.load(recv);
+    let flags_opnd = Opnd::mem(VALUE_BITS, recv_reg, RUBY_OFFSET_RBASIC_FLAGS);
+    let encoding_flags_opnd = asm.and(flags_opnd, Opnd::UImm(RUBY_ENCODING_MASK as u64));
+    let encoding_index = asm.rshift(encoding_flags_opnd, Opnd::UImm(RUBY_ENCODING_SHIFT as u64));
+    asm.cmp(encoding_index, Opnd::UImm(RUBY_ENCINDEX_ASCII_8BIT as u64));
+    asm.jne(generic_str_concat_codepoint);
+
+    // Check if the codepoint is limited to a byte value. If it isn't, we need to use the generic
+    // string concatenation, which will ultimately raise a `RangeError` as ASCII-8BIT strings only
+    // accept codepoint values that fit in a byte range.
+    asm_comment!(asm, "Check if codepoint is a byte value");
+    let codepoint_untag = asm.rshift(codepoint, Opnd::UImm(1));
+    asm.cmp(codepoint_untag, Opnd::Imm(0xff));
+    asm.jg(generic_str_concat_codepoint);
+
+    // If we've made it this far, we must be appending a single byte value to a binary string.
+    asm_comment!(asm, "Optimized (string, codepoint) concatenation");
+    let ret_opnd = asm.ccall(rb_str_buf_cat_byte as *const u8, vec![recv, codepoint_untag]);
+    let stack_ret = asm.stack_push(Type::TString);
+    asm.mov(stack_ret, ret_opnd);
+    asm.stack_pop(1);
```

Comment:
Noticing the `stack_pop` here. One thing to know is that `asm.stack_pop` is not going to care whether the conditional branches are taken or not.

Reply:
I guess just be aware that the stack_pop is always executed, not conditionally executed.